Migrate to header-only, involved generic abstract base class for data holder class, etc...

Gluttton opened the pull request:
I'd like to propose set of my changes (some of them are very specific and subjective) as one bunch.

**List of the most valuable changes**.
**Major**.
1. Extracted all necessary functions which required for navigation and access and put into generic abstract base class. This allows rewrite existed adapters and simplify creation new adapters. Added adapters for Eigen, OpenCV, Armadillo and other matrices (#6).
2. Also this allows eliminate unnecessary copying of input matrix into inner matrix. And remove inner matrix member for storage input data.
3. Fix error for integral types (calling of replace_infinities function).
4. Made algorithm tolerant to unsigned types (as consequence change the return value: 0 - is still means solution, but instead of -1 is used +1).

**Medium**.
1. Put source files into subdirectory to prevent file names pollution.
2. Put source code into namespace to prevent name pollution.
3. Added set of examples which step-by-step shows usage of the library.
4. Existed tests of the algorithm are parametrized by type.
5. Eliminate function ```pair_in_list``` and use ```std::find``` instead.

**Minor**.
1. Migrate to header-only library.
2. Added primitive macro which emulate Design by Contract directives (#5).
3. Added beautifier rules (#16).

For more details see: https://gitlab.com/Gluttton/munkres-cpp .

P.S. Sorry for my English.
.

saebyn commented:
Thanks! I see a few typos in the README, and it looks like there are some conflicts before this can be merged. I'll have a closer look later on this week.

Gluttton commented:
> I see a few typos in the README

Thanks! Examples may be have errors too.

> I'll have a closer look later on this week.

It's OK. Please carefully check all changes, may be some of them isn't good enough (especially minor 2 and 3).

> it looks like there are some conflicts before this can be merged

Yep. One possible approach (which I'm going to apply) just rewrite.

saebyn commented:
I get an error when linking the tests, it seems to want to link in Qt5::Gui.

```
[100%] Linking CXX executable munkrestest
/usr/bin/ld: cannot find -lQt5::Gui
collect2: error: ld returned 1 exit status
tests/CMakeFiles/munkrestest.dir/build.make:174: recipe for target 'tests/munkrestest' failed
make[3]: *** [tests/munkrestest] Error 1
CMakeFiles/Makefile2:544: recipe for target 'tests/CMakeFiles/munkrestest.dir/all' failed
make[2]: *** [tests/CMakeFiles/munkrestest.dir/all] Error 2
CMakeFiles/Makefile2:583: recipe for target 'tests/CMakeFiles/tests.dir/rule' failed
make[1]: *** [tests/CMakeFiles/tests.dir/rule] Error 2
Makefile:357: recipe for target 'tests' failed
make: *** [tests] Error 2
```

saebyn commented on README.md:
```diff
@@ -1,8 +1,8 @@
 
-[![build](https://travis-ci.org/saebyn/munkres-cpp.svg?branch=master)](https://travis-ci.org/saebyn/munkres-cpp)
-[![codecov.io](http://codecov.io/github/saebyn/munkres-cpp/coverage.svg?branch=master)](http://codecov.io/github/saebyn/munkres-cpp?branch=master)
+[![build status](https://gitlab.com/Gluttton/munkres-cpp/badges/devel/build.svg)](https://gitlab.com/Gluttton/munkres-cpp/commits/devel)
```
Did you intend to include these changes?

saebyn commented on README.md:
```diff
@@ -31,7 +31,8 @@ For development:
  - microbenchmaring requires [Benchmark](https://github.com/google/benchmark), [Celero](https://github.com/DigitalInBlue/Celero), [Hayai](https://github.com/nickbruun/hayai) and [gprof](http://www.gnu.org/software/binutils/);  
  - code coverage requires [gcov](https://gcc.gnu.org/onlinedocs/gcc/Gcov.html) and lcov;  
- - static code analyzis required [cppcheck](https://github.com/danmar/cppcheck).  
+ - static code analyzis required [cppcheck](https://github.com/danmar/cppcheck);  
```
"analyzis required" should be "analysis requires"

Gluttton commented:
> I get an error when linking the tests, it seems to want to link in Qt5::Gui.

This is error. When some package is missing (Qt, OpenCV, Boost, etc.) it should silently skip such test. I will fix it.

For now can you just remove the following line from `tests/CMakeLists.txt`:  
```
target_link_libraries (
    munkrestest
    ${GTEST_BOTH_LIBRARIES}
    ${OpenCV_LIBS}
    Qt5::Gui        <-- Remove this line.
    gcov
    pthread
)
```

Gluttton commented:
> "analyzis required" should be "analysis requires"

Fixed.

> I get an error when linking the tests, it seems to want to link in Qt5::Gui.

Fixed.

saebyn commented:
Looking good

Gluttton commented:
> Did you intend to include these changes?

There are two possible ways. Continue use Travic-CI and migrate to GitLab-CI. For me configure Travic-CI is very difficult. I spent a lot of time when I configured build job for our previous version. For current version I have no idea how to configure build job which include Eigen, OpenCV, etc... Also a few days ago builds on Travis-CI were fails because "gcov-4.8 not found". I think it was some problems with ppa:ubuntu-toolchain-r-test which we use.

Instead!

Last version of GitLab-CI allows use custom (!) Docker (!) images! I've created [image](https://hub.docker.com/r/gluttton/munkres-cpp/) which allows build and test everything (tests, benchmarks and examples) in the project.

So now is a question: which CI server to use :-)  .

P.S. Do you have GitLab account?

Gluttton commented:
> Looking good

@saebyn , does it mean that the request can be merged?

saebyn commented:
This still has conflicts that prevent me from merging it.

Gluttton commented:
I try to fix the problem in a dirty way (commit all changes as one). Let me know if this is acceptable or not.

Gluttton commented:
Have you seen my last message (about month ago)?

saebyn commented:
This introduces many spelling and grammar errors in the README, and its a rather too large set of changes for me to review in detail in one go. I would prefer seeing separate pull requests for individual items ("Added beautifier rules",  "Eliminate function pair_in_list and use std::find instead") - even if some of them were built on top of other pull requests. I don't see anything wrong with the code from my read through, but honestly, I don't have the time/motivation to be more active on this project right now.

Gluttton commented:
@saebyn , thanks for your answer! I understood you.

>I would prefer seeing separate pull requests for individual items

I will create separate pull requests for the most important changes.

> I don't have the time/motivation to be more active on this project right now

This is very sad.
